lib/ex_machina/ecto_strategy.ex

```diff
@@ -70,40 +59,76 @@ defmodule ExMachina.EctoStrategy do
     end
   end
 
-  defp cast_all_assocs(%{__struct__: schema} = struct) do
-    assocs = get_schema_assocs(schema)
+  defp cast_all_embeds(%{__struct__: schema} = struct) do
+    embed_keys = schema_embeds(schema)
 
-    Enum.reduce(assocs, struct, fn(assoc, struct) ->
-      casted_value = struct |> Map.get(assoc) |> cast_assoc(assoc, struct)
+    Enum.reduce(embed_keys, struct, fn(embed_key, struct) ->
+      casted_value = Map.get(struct, embed_key) |> cast_embed(embed_key, struct)
 
-      Map.put(struct, assoc, casted_value)
+      Map.put(struct, embed_key, casted_value)
     end)
   end
 
-  defp cast_assoc(original_assoc, assoc_key, %{__struct__: schema} = struct) do
-    case original_assoc do
-      has_or_embeds_many when is_list(has_or_embeds_many) ->
-        Enum.map(has_or_embeds_many, &(cast_assoc(&1, assoc_key, struct)))
+  defp cast_embed(embeds_many, embed_key, struct) when is_list(embeds_many) do
+    Enum.map(embeds_many, &(cast_embed(&1, embed_key, struct)))
+  end
+  defp cast_embed(embed, embed_key, %{__struct__: schema}) do
+    case embed do
+      %{} ->
+        embedding_reflection = schema.__schema__(:embed, embed_key)
+        embed_type = embedding_reflection.related
+        struct(embed_type) |> Map.merge(embed) |> cast()
+
+      nil -> nil
+    end
+  end
+
+  defp cast_all_assocs(%{__struct__: schema} = struct) do
+    assoc_keys = schema_associations(schema)
+
+    Enum.reduce(assoc_keys, struct, fn(assoc_key, struct) ->
+      casted_value = Map.get(struct, assoc_key) |> cast_assoc(assoc_key, struct)
+
+      Map.put(struct, assoc_key, casted_value)
+    end)
+  end
 
+  defp cast_assoc(has_many_assoc, assoc_key, struct) when is_list(has_many_assoc) do
+    Enum.map(has_many_assoc, &(cast_assoc(&1, assoc_key, struct)))
+  end
+  defp cast_assoc(assoc, assoc_key, %{__struct__: schema}) do
+    case assoc do
       %{__meta__: %{__struct__: Ecto.Schema.Metadata, state: :built}} ->
-        cast(original_assoc)
+        cast(assoc)
 
       %{__struct__: Ecto.Association.NotLoaded} ->
-        original_assoc
+        assoc
 
       %{__struct__: _} ->
-        cast(original_assoc)
+        cast(assoc)
 
       %{} ->
-        assoc_reflection = schema.__schema__(:association, assoc_key) || schema.__schema__(:embed, assoc_key)
+        assoc_reflection = schema.__schema__(:association, assoc_key)
         assoc_type = assoc_reflection.related
-        assoc_type |> struct() |> Map.merge(original_assoc) |> cast()
+        struct(assoc_type) |> Map.merge(assoc) |> cast()
 
       nil -> nil
     end
   end
 
-  defp get_schema_assocs(schema) do
-    schema.__schema__(:associations) ++ schema.__schema__(:embeds)
+  defp schema_fields(schema) do
+    schema_non_virtual_fields(schema)  -- schema_embeds(schema)
```

Comment:
extra space here before `--`

Reply:
Nice catch. Updated in b2ec4ca002be6d5a91609b45a4582e4652573e8d